opnode/node/server.go

```diff
@@ -0,0 +1,257 @@
+package node
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/ethereum-optimism/optimistic-specs/opnode/l2"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/holiman/uint256"
+	"github.com/protolambda/zssz"
+)
+
+// TODO(inphi): add metrics
+
+var (
+	ErrParseErr = &RPCErr{
+		Code:          -32700,
+		Message:       "parse error",
+		HTTPErrorCode: 400,
+	}
+	ErrInternal = &RPCErr{
+		Code:          JSONRPCErrorInternal,
+		Message:       "internal error",
+		HTTPErrorCode: 500,
+	}
+)
+
+func ErrInvalidRequest(msg string) *RPCErr {
+	return &RPCErr{
+		Code:          -32601,
+		Message:       msg,
+		HTTPErrorCode: 400,
+	}
+}
+
+type RPCHandler func(context.Context, *RPCReq) *RPCRes
+
+type rpcServer struct {
+	l2RPCClient            *rpc.Client
+	server                 *http.Server
+	withdrawalContractAddr common.Address
+	log                    log.Logger
+	appVersion             string
+}
+
+func newRPCServer(ctx context.Context, addr string, port int, l2RPCClient *rpc.Client, withdrawalContractAddress common.Address, log log.Logger, appVersionn string) *rpcServer {
+	r := &rpcServer{
+		l2RPCClient:            l2RPCClient,
+		withdrawalContractAddr: withdrawalContractAddress,
+		log:                    log,
+		appVersion:             appVersionn,
+	}
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/", r.rootHandler)
+	mux.HandleFunc("/healthz", r.healthzHandler)
+	addrPort := fmt.Sprintf("%s:%d", addr, port)
+	r.server = &http.Server{
+		Handler: mux,
+		Addr:    addrPort,
+	}
+
+	return r
+}
+
+func (s *rpcServer) Start() {
+	go func() {
+		if err := s.server.ListenAndServe(); err != nil {
+			if errors.Is(err, http.ErrServerClosed) {
+				s.log.Info("RPC server shutdown")
+				return
+			}
+			s.log.Error("error starting RPC server", "err", err)
+		}
+	}()
+}
+
+func (r *rpcServer) Stop() {
+	_ = r.server.Shutdown(context.Background())
+}
+
+func (s *rpcServer) rootHandler(w http.ResponseWriter, r *http.Request) {
+	s.log.Info("received RPC request", "user_agent", r.Header.Get("user-agent"))
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		s.log.Error("error reading request body", "err", err)
+		writeRPCError(s.log, w, nil, ErrInternal)
+		return
+	}
+
+	req, err := ParseRPCReq(body)
+	if err != nil {
+		s.log.Info("error parsing RPC call", "err", err)
+		writeRPCError(s.log, w, nil, err)
+		return
+	}
+
+	if err := ValidateRPCReq(req); err != nil {
+		writeRPCRes(s.log, w, NewRPCErrorRes(nil, err))
+		return
+	}
+
+	if req.Method != "optimism_outputAtBlock" {
```

Comment:
Any opinions on using the rpc package from `go-ethereum`?

https://github.com/ethereum/go-ethereum/blob/a8040bc2c51605a4cca9e48cac83ff14928d50c2/node/rpcstack.go#L284

Reply:
A few reasons why I think this is a good idea. It makes us become more familiar with the geth codebase, it makes writing RPC handlers really easy, its very hardened and it makes it easy for contributors familiar with geth be familiar with our codebase. I'm not 100% sure it can be easily used standalone but I'm pretty sure that it can be

Reply:
I have no objections to using the package. I'll see if that can be used.

Reply:
If we can, definitely just import things from the go-ethereum `rpc` package. We already use the client-side of it in other places, and we don't have to roll our own (cool stuff though)